### PR TITLE
feat(builder): add config normalizer

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Description

Add the configuration normalizer, but not force the plugins to use it for now.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/builder-webpack-provider/src/core/initPlugins.ts [hunk 1]
```diff
@@ -18,10 +18,17 @@ export async function initPlugins({
   const publicContext = createPublicContext(context);
 
   const getBuilderConfig = () => context.config;
+  const getNormalizedConfig = () => {
+    if (!context.normalizedConfig) {
+      throw new TypeError('Normalized config is not ready.');
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/core/initPlugins.ts:23]: 为啥是 TypeError

File: packages/builder/builder-webpack-provider/src/config/defaults.ts [hunk 2]
```diff
@@ -8,61 +8,82 @@ import {
...
-      media: MEDIA_DIST_DIR,
+const defineDefaultConfig = extendsType<BuilderConfig>();
+
+export const createDefaultConfig = () =>
+  defineDefaultConfig({
+    dev: {
+      port: 3000,
+      assetPrefix: '/',
+      https: false,
+      startUrl: true,
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/config/defaults.ts:23]: startUrl 的默认值不是 true

File: packages/builder/builder-webpack-provider/src/config/defaults.ts [hunk 3]
```diff
@@ -8,61 +8,82 @@ import {
...
-      svg: SVG_DIST_DIR,
-      font: FONT_DIST_DIR,
-      image: IMAGE_DIST_DIR,
-      media: MEDIA_DIST_DIR,
+const defineDefaultConfig = extendsType<BuilderConfig>();
+
+export const createDefaultConfig = () =>
+  defineDefaultConfig({
+    dev: {
+      port: 3000,
```
Comment (chenjiahan) [inline packages/builder/builder-webpack-provider/src/config/defaults.ts:20]: port 的默认值不是 8080 吗

File: packages/builder/builder-webpack-provider/src/config/defaults.ts [hunk 4]
```diff
@@ -8,61 +8,82 @@ import {
...
+      hmr: true,
+      progressBar: true,
+    },
+    html: {
+      crossorigin: false,
+      disableHtmlFolder: false,
+    },
+    tools: {
+      tsChecker: {},
+      // babel: {},  -> should be closable explicitly
```
[the remark this answers is not in the document]
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/config/defaults.ts:33]: 之前这里会填充空对象作为 babel 的默认值，但是这样会改变 babel 的默认行为，考虑之后把这块类型改一下如何：

```typescript
type ClosableConfig<T> = NonNullable<T> | boolean;
const config = getNormalizedConfig();
config.tools.babel
//                   ^? ToolsBabelConfig | undefined -> ClosableConfig<ToolsBabelConfig>
```

File: packages/builder/builder-webpack-provider/src/config/defaults.ts [hunk 5]
```diff
@@ -8,61 +8,82 @@ import {
...
+      devServer: {
+        client: {},
+        devMiddleware: {
+          writeToDisk: true,
```
[the remark this answers is not in the document]
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/config/defaults.ts:37]: 确实，这个属于是独占的配置，我给删掉

File: packages/builder/builder-webpack-provider/src/core/initPlugins.ts (hunk 1 above)
Comment (Asuka109) [reply packages/builder/builder-webpack-provider/src/core/initPlugins.ts:23]: 随手写的🤔这里要改么

Comment (chenjiahan) [reply packages/builder/builder-webpack-provider/src/core/initPlugins.ts:23]: 可以改一下，因为不符合 TypeError 的语义，用 Error 就可以咯

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**